onnxruntime/core/optimizer/double_qdq_pairs_remover.cc

```diff
@@ -175,16 +81,141 @@ bool DoubleQDQPairsRemover::FindNewZeroPointAndScale(const Graph& graph, const N
   return true;
 }
 
-template <typename T>
-void DoubleQDQPairsRemover::ApplyNewInputValue(Graph& graph, Node& node, const InputIndex& index, T value) {
-  const auto* input_tensor = graph_utils::GetConstantInitializer(graph, node.InputDefs()[index]->Name());
-  Initializer input_init{*input_tensor, graph.ModelPath()};
-  TensorProto new_input_tensor(*input_tensor);
-  input_init.data<T>()[0] = value;
-  input_init.ToProto(new_input_tensor);
-  auto new_name = graph.GenerateNodeArgName("DoubleQDQRemoved_" + node.InputDefs()[index]->Name());
-  new_input_tensor.set_name(new_name);
-  NodeArg& new_input = graph_utils::AddInitializer(graph, new_input_tensor);
-  graph_utils::ReplaceNodeInput(node, index, new_input);
+// Recomputes the zero point and scale of the outer Q/DQ nodes (i.e., Q1 and DQ2). This is necessary because
+// the original two QDQ pairs may have different zero-points and scales. Ex: Q1 -> DQ1 -> Q2 -> DQ2, where
+// the first pair has (zp1, scale1) and the second pair has (zp2, scale2).
+// After removing the middle two nodes, the zero point and scale of the final (outer) ops must be recomputed
+// for correctness.
+template <typename ZeroPointType>
+static bool RecomputeOuterQDQZeroPointAndScale(Graph& graph, const Node& dq1, const Node& q2, Node& q1, Node& dq2) {
```

Comment:
nit: the ordering of the inputs here and in IsReducibleDoubleQDQSequence is a little weird now that they're renamed.

Reply:
Updated